account_asset_maintenance/models/account_invoice.py

```diff
@@ -0,0 +1,68 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 Tecnativa - Pedro M. Baeza
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models
+import math
+
+
+class AccountInvoice(models.Model):
+    _inherit = 'account.invoice'
+
+    equipment_ids = fields.Many2many(
+        comodel_name="maintenance.equipment", compute="_compute_equipment_ids",
+        string="Equipments",
+    )
+
+    @api.multi
+    @api.depends('invoice_line_ids', 'invoice_line_ids.equipment_ids')
+    def _compute_equipment_ids(self):
+        for invoice in self:
+            invoice.equipment_ids = [
+                (6, 0, invoice.mapped('invoice_line_ids.equipment_ids').ids),
+            ]
+
+    @api.multi
+    def action_invoice_cancel(self):
+        res = super(AccountInvoice, self).action_invoice_cancel()
+        self.mapped('equipment_ids').unlink()
+        return res
+
+
+class AccountInvoiceLine(models.Model):
+    _inherit = 'account.invoice.line'
+
+    equipment_ids = fields.One2many(
+        comodel_name="maintenance.equipment", inverse_name="invoice_line_id",
+        string="Equipments",
+    )
+
+    def _prepare_equipment_vals_list(self, invoice_line):
+        vals_list = []
+        num = int(math.ceil(invoice_line.quantity))
+        for i in range(num):
+            vals_list.append({
+                'name': "{} [{}/{}]".format(invoice_line.name, i + 1, num),
+                'category_id': (
+                    invoice_line.asset_category_id.equipment_category_id.id
+                ),
+                'invoice_line_id': invoice_line.id,
+                'cost': invoice_line.price_subtotal / invoice_line.quantity,
+                'partner_id': invoice_line.invoice_id.partner_id.id,
+            })
+        return vals_list
+
+    @api.multi
+    def asset_create(self):
+        for line in self.filtered('asset_category_id.equipment_category_id'):
+            # Create equipments
+            equipments = self.env['maintenance.equipment']
+            for vals in self._prepare_equipment_vals_list(line):
+                equipments += equipments.create(vals)
```

Comment:
@pedrobaeza in my humble opinion you have to make several modifications to the code, but then it would fulfill what you propose and also what comments @dreispt. Among the modifications should be add a check mark in the category of the Asset which are Equipment. So we could buy an Apartment and enter as an Asset and not as a Equipment. You should create a relationship between the Category of the Equipment and the Category of the Asset for when you want to register a Equipment (which because of the inheritance is already created as an Asset) that I have not purchased. Other changes would have to be made, but I think the module would be a bit more complete.